### ISSUE TITLE
SRIOV CNI to determine the driver and user config 

### ISSUE THREAD

**Issue by pkdas** [intel/sriov-cni#58]
Currently the kernel driver is specified as part of SRIOV CNI config. And the SRIOV plugin binds to the user specified driver if L2Enable is True. If DPDK is enabled, runs dpdk_devbind to bind to a dpdk driver. 
       
spec:
  config: '{
    "cniVersion": "0.3.0",
    "type": "sriov",
    "vlan": 200,
    "dpdk": {
      "kernel_driver":"i40evf",
      "dpdk_driver":"igb_uio",
      "dpdk_tool":"/usr/local/sbin/dpdk-devbind"
    }
  }'

This config does not work if there are multiple different types of NIC (i40e, ConnectX)
This means that one requires network definition based on NIC type. 

Currently L2Enable and DPDK are mutually exclusive. But the ConnectX support in DPDK does not require dpdk-devbind to be called and instead it requires that the VF/PF binds to the kernel driver. As a result one needs multiple definitions of a network based on NIC types even though the DPDK application in a POD consumes both the NIC types.

Instead the SRIOV plugin can itself determine the driver type based on VF/PF PCIeaddr.

I believe the NIC type, driver to bind/unbind belongs to SRIOV plugin and will allow a generic network definition which is defined by SRIOVDP and the resource type.

Basically remove dependency on      
 "kernel_driver":"i40evf", 
"dpdk_driver":"igb_uio", 

and let SRIOV plugin determine the driver based on VF/PF/pcie-addr/sysfs.

This should work for both cases with and without SRIOVDP.

**Comment by pkdas** [comment]
Multiple Network definition based on NIC types use-cases.

1) "kernel_driver" is different because NIC (vendor/device) is different 
2) "L2Enable" True because NIC type (ConnectX) does not require dpdk-devbind to be called but enables DPDK in the Container application

**Comment by moshe010** [comment]
@pkdas, Can you share the container yaml that you use? and How do you run the testpmd to test it?Mellanox Card. for the dpdk I think it should be
"kernel_driver":"mlx5_core",
"dpdk_driver":"mlxr_core"
basically we shouldn't do any driver change as the same driver work for kernel and dpdk. 

Regrading to support multi nic, I think we will depreciate this option in the SR-IOV cni and move it to the SR-IOV device plugin.

**Comment by zshi-redhat** [comment]
> Currently the kernel driver is specified as part of SRIOV CNI config. And the SRIOV plugin binds to the user specified driver if L2Enable is True. If DPDK is enabled, runs dpdk_devbind to bind to a dpdk driver.
> 
> spec:
> config: '{
> "cniVersion": "0.3.0",
> "type": "sriov",
> "vlan": 200,
> "dpdk": {
> "kernel_driver":"i40evf",
> "dpdk_driver":"igb_uio",
> "dpdk_tool":"/usr/local/sbin/dpdk-devbind"
> }
> }'
> 
> This config does not work if there are multiple different types of NIC (i40e, ConnectX)
> This means that one requires network definition based on NIC type.
> 
> Currently L2Enable and DPDK are mutually exclusive. But the ConnectX support in DPDK does not require dpdk-devbind to be called and instead it requires that the VF/PF binds to the kernel driver. As a result one needs multiple definitions of a network based on NIC types even though the DPDK application in a POD consumes both the NIC types.
> 
> Instead the SRIOV plugin can itself determine the driver type based on VF/PF PCIeaddr.
> 
> I believe the NIC type, driver to bind/unbind belongs to SRIOV plugin and will allow a generic network definition which is defined by SRIOVDP and the resource type.
> 
> Basically remove dependency on
> "kernel_driver":"i40evf",
> "dpdk_driver":"igb_uio",
> 
> and let SRIOV plugin determine the driver based on VF/PF/pcie-addr/sysfs.

I assume by saying SRIOV plugin, you mean SRIOV CNI.
There are things that SRIOV CNI cannot gracefully handle, for example, after device binding to vfio-pci, SRIOV CNI cannot instruct device spec(/dev/vfio/vfio, /dev/vfio/<iommu_group>) mounting to Pod (but device plugin API supports passing mount directory to container).

> 
> This should work for both cases with and without SRIOVDP.

**Comment by killianmuldoon** [comment]
Closing as stale.
